File: doublyLinkedList.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None
        self.prev = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.curr = None 
        self.tail = None

        
    def addEnd(self,data):
        new = Node(data)
        if self.head is None:
            self.head = new
            self.tail= new
            self.head.prev=self.tail
            self.tail.next=self.head
        
        curr = self.tail
        new.prev=self.tail
        self.tail.next=new
        self.tail=new
        self.tail.next=self.head
        self.head.prev=self.tail
        
        
    def addStart(self,data):
        new = Node(data)
        if self.head is None:
            self.head = new
            self.tail= new
            self.head.prev=self.tail
            self.tail.next=self.head
        
        curr = self.head
        new.next=self.head
        self.head.prev=new
        self.head=new
        self.head.prev=self.tail
        self.tail.next = self.head
        
        
    def popStart(self):
        if self.head is None:
            raise IndexError("list is empty")
        
        curr = self.head.next
        self.head = curr
        self.head.prev=self.tail
        self.tail.next = self.head
        
    def popEnd(self):
        if self.head is None:
            raise IndexError("list is empty")
        
        curr = self.tail.prev
        self.tail = curr
        self.tail.next=self.head
        self.head.prev = self.tail
        
        

    def insert(self,data,i):
        new = Node(data)
        counter = 0
        if self.head is None:
            self.head = new
            self.tail= new
            self.head.prev=self.tail
            self.tail.next=self.head
            return
        
        curr = self.head
        while curr.next!=self.head and counter < i:
            curr = curr.next
            counter+=1
        if curr is self.head:
            self.addStart(data)
        elif curr is self.tail:
            self.addEnd(data)
        else:
            prevNode = curr.prev
            prevNode.next=new
            new.prev=prevNode
            curr.prev=new
            new.next=curr
            
            
    def delete(self,i):
        counter = 0
        if self.head is None:
            raise IndexError("list is empty")
        
        curr = self.head
        while curr.next!=self.head and counter < i:
            curr = curr.next
            counter+=1
        if curr is self.head:
            self.popStart()
        elif curr is self.tail:
            self.popEnd()
        else:
            prevNode=curr.prev
            nextNode=curr.next
            prevNode.next=nextNode
            nextNode.prev=prevNode
            
    def __iter__(self):
        return self
        
    def __next__(self):
        if self.head and not self.curr:
            self.curr = self.head
            return self.curr.data
        try:
            if self.curr.next!=self.head:
                self.curr = self.curr.next
                return self.curr.data
        except AttributeError:
            raise StopIteration("list is empty")

            
        else:
            self.curr=None
            raise StopIteration
        

    def __str__(self):
        curr = self.head
        table=[]
        while curr:
            #print(curr.prev.data,curr.data,curr.next.data)
            table.append(curr.data)
            curr = curr.next
            if curr == self.head:
                return str(table)

                
    
    def __repr__(self):
        return self.__str__()


if __name__ == "__main__": 
    s = LinkedList()
    s.addEnd(1)
    print(s)

    s.addEnd(2)
    print(s)

    s.addEnd(3)
    print(s)

    s.addStart(4)
    print(s)

    s.addStart(5)
    print(s)

    s.addStart(6)
    print(s)
    
    s.insert(22,1)
    print(s)
    
    s.delete(6)
    print(s)
    
    s.popStart()
    print(s)
    
    s.popStart()
    print(s)
    
    s.popEnd()
    print(s)

    print("--------------------------")
    
    for i in s:
        print(i)
        
    print("--------------------------")
        
#     kk=LinkedList()
#     for i in kk:
#         print(i)
        
#     print("--------------------------")
        
#     m=LinkedList()
#     m.popEnd()
    
    
#     print("--------------------------")
        
#     m=LinkedList()
#     m.popStart()
    
        
    

File: DoublyLinkedList.java
package linkedList;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class DoublyLinkedList<T> {
	
	Node<T> head;
	Node<T> tail;

	
	class Node<T>
	{
		private T data = null;
		Node<T> next;
		Node<T> prev;
		Node(T data)
		{
			this.data = data;
			this.next=null;
			this.prev=null;
			
		}
		
		public T getData()
		{
			return data;
		}
		
		void setData(T data)
		{
			this.data=data;
		}
		
		
	}
	
	public DoublyLinkedList()
	{
		this.head=null;
		this.tail=null;
	}
	
	@SuppressWarnings("unchecked")
	public String toString()
	{
		Node curr = head;
		List<T> object = new ArrayList<T>();
		
		if(curr == null)
		{
			System.out.println("Empty List");
		}
		
		else
		{
			while(curr!=null)
			{
			object.add((T) curr.getData());
			curr = curr.next;
			if(curr==head)
			{
				return  Arrays.toString(object.toArray());
			}
			
			
			}
		}
		
		return  Arrays.toString(object.toArray());
	}
	
	
	public void pushFirst(T value)
	{
		Node element = new Node(value);
		
		if(head==null)
		{
			head = element;
		    tail = element;
		    head.prev=tail;
		    tail.next=head;
		    return;
		}
		
		Node curr = head;
		element.next=head;
		head.prev=element;
		head=element;
		head.prev=tail;
		tail.next = head;
		
	}
	
	
	public void pushEnd(T value)
	{
		Node element = new Node(value);
		
		if(head==null)
		{
			head = element;
		    tail = element;
		    head.prev=tail;
		    tail.next=head;
		}
        
		Node curr = tail;
		element.prev=tail;
        tail.next=element;
        tail=element;
        tail.next=head;
        head.prev=tail;
	}
	
	public T popFirst() throws EmptyListException 
	{
        if(head == null)
        {
        	throw new EmptyListException("EmptyList");
        }
        else if(tail==head)
        {
            T val = head.getData();
            this.head=null;
            this.tail=null;
            return val;
        }
        
        T val = head.getData();
        Node curr = head.next;
        this.head = curr;
        this.head.prev=this.tail;
        this.tail.next = this.head;
        return val;
	}
	
	public T popEnd() throws EmptyListException  
	{
        if(head == null)
        {
        	throw new EmptyListException("EmptyList");
        }
        else if(tail==head)
        {
            T val = head.getData();
            this.head=null;
            this.tail=null;
            return val;
        }
        
        T val = tail.getData();
        Node curr = tail.prev;
        this.tail = curr;
        this.tail.next=this.head;
        this.head.prev = this.tail;
        return val;
	}
	
	
	public T deleteByIndex(int id) throws EmptyListException
	{
		int counter = 0;
		
		if(this.head == null)
        {
        	throw new EmptyListException("EmptyList");
        }
		
		else if(id == counter)
	    {
	         return this.popFirst();
	    }
		else
		{
		 Node curr = this.head.next;
		 while(curr!=head)
		 {
			 counter++;
			 if(id==counter)
			 {
				 if(curr==tail)
	              {
	                  return popEnd();
	              }
				 else
				 {
				 T val = (T) curr.getData();
		         Node prevNode=curr.prev;
		         Node nextNode=curr.next;
		         prevNode.next=nextNode;
		         nextNode.prev=prevNode;
		         return val; 
				 }
			 }
			 curr = curr.next;
		 }
		}
		 
		 throw new IndexOutOfBoundsException("index max = "+counter+" your index is "+id); 
	}
	
	
	public void insertByIndex(int id, T value) throws EmptyListException
	{
		Node element = new Node(value);
		int counter = 0;
		
		if(head==null)
		{
			head = element;
		    tail = element;
		    head.prev=tail;
		    tail.next=head;
		    return;
		}
		
		else if(id == counter)
		{
	        this.pushFirst(value);
			return;   
	    }
		else
		{
		 Node curr = this.head.next;
		 while(curr!=head)
		 {
			 counter++;
			 if(id==counter)
			 {
				 if(curr==tail)
	              {
					  pushEnd(value);
	                  return;
	              }
				 else
				 {
				 Node prevNode = curr.prev;
				 prevNode.next=element;
				 element.prev=prevNode;
				 curr.prev=element;
				 element.next=curr;
		         return; 
				 }
			 }
			 curr = curr.next;
		 }
		}
		
		 throw new IndexOutOfBoundsException("index max = "+counter+" your index is "+id); 
		 
	}
	
	
	public static void main(String[] args) throws EmptyListException
	{
		DoublyLinkedList l = new DoublyLinkedList();
		l.pushFirst("ala");
		l.pushFirst("ma");
		l.pushFirst("Kota");
		l.pushEnd("xd");
		System.out.println(l);
		l.popEnd();
		l.popFirst();
		System.out.println(l);
		l.popEnd();
		l.popFirst();
		System.out.println(l);
		//l.popStart();
		l.pushEnd("ala");
		l.pushEnd("ma");
		l.pushEnd("Kota");
		System.out.println(l);
		l.deleteByIndex(2);
		System.out.println(l);
		l.pushEnd("Kota");
		System.out.println(l);
		l.insertByIndex(0, "nr 0");
		l.insertByIndex(3, "nr 3");
		l.insertByIndex(2, "nr 1");
		System.out.println(l);
		for(int i = 5; i>=0; i--)
		{
			l.deleteByIndex(i);
			System.out.println(l);

		}



		





	}

}

class EmptyListException extends Exception {
    public EmptyListException(String message) {
        super(message);
    }
}
